### PR TITLE
Issue72

### PR DESCRIPTION
**Pull request by ClemensBuechner**

### CODE REVIEW

File: machetli/environments.py [hunk 1]
```diff
@@ -357,106 +359,94 @@ def _wait_for_filesystem(self, *paths):
...
-                raise PollingError(job_id)
-
-    def _build_task_state_dict(self, sacct_output):
-        unclean_job_state_list = sacct_output.strip("\n").split("\n")
-        stripped_job_state_list = [pair.strip(
-            "+ ") for pair in unclean_job_state_list]
-        return {k: v for k, v in
-                (pair.split() for pair in stripped_job_state_list)}
+                    task.status = EvaluationTask.CRITICAL
+                    task.error = f"Unexpected exit code {exit_code}"
```
Comment (ClemensBuechner) [inline machetli/environments.py:441]:
```suggestion
                    task.error = f"Unexpected evaluator exit code {exit_code}"
```

File: machetli/search.py [hunk 2]
```diff
@@ -1,8 +1,10 @@
 import logging
+import time
```
Comment (ClemensBuechner) [inline machetli/search.py:2]: `time` appears to be unused.

File: machetli/environments.py [hunk 3]
```diff
@@ -63,41 +80,38 @@ def __init__(self, allow_nondeterministic_successor_choice=True,
         self.allow_nondeterministic_successor_choice = \
             allow_nondeterministic_successor_choice
 
-    def submit(self, batch, evaluator_path):
+    def run(self, evaluator_path, batch, on_task_completed):
```
Comment (ClemensBuechner) [inline machetli/environments.py:83]: We expect this function to have a return value in the search, so we should proably ensure this is the case.
```suggestion
    def run(self, evaluator_path, batch, on_task_completed) -> list[EvaluationTask]:
```

File: machetli/search.py [hunk 4]
```diff
@@ -86,18 +88,62 @@ def search(initial_state, successor_generator, evaluator_path, environment=None)
...
+
+def _get_improving_successor(evaluator_path, successors, environment):
+    report_if_no_improvement = set()
+    def on_task_completed(job_id, task):
+        if environment.allow_nondeterministic_successor_choice:
+            # If an evaluation reports a non-improving successor,
+            # deterministic mode considers the next successor. In all
+            # other cases (improving successor, evaluation did not
+            # finish correctly), there is no need to evaluate later
+            # successors.
```
Comment (ClemensBuechner) [inline machetli/search.py:115]: The comment above seems a bit redundant with the code expressing pretty much the same.

File: machetli/environments.py [hunk 5]
```diff
@@ -63,41 +80,38 @@ def __init__(self, allow_nondeterministic_successor_choice=True,
         self.allow_nondeterministic_successor_choice = \
```
Comment (ClemensBuechner) [inline machetli/environments.py:80]: I think this should no longer belong to the `Environment`; it is only used by the search.

File: machetli/search.py [hunk 6]
```diff
@@ -86,18 +88,62 @@ def search(initial_state, successor_generator, evaluator_path, environment=None)
...
+            if task.status == EvaluationTask.DONE_AND_NOT_IMPROVING:
+                continue
+            elif task.status == EvaluationTask.DONE_AND_IMPROVING:
+                return task.successor.state, task.successor.change_msg
+            elif not environment.allow_nondeterministic_successor_choice:
+                return None, (
+                    "Errors occurred in successor evaluations before finding "
+                    "an improving successor. With the option "
+                    "`allow_nondeterministic_successor_choice` an improving "
+                    "successor found later would not count.")
```
Comment (ClemensBuechner) [inline machetli/search.py:140]: Should we report here what kind of error occurred? Or do we do so somewhere else? Also, the error message confuses me: With the option, we *would* consider a successor found later, wouldn't we?

File: machetli/search.py [hunk 7]
```diff
@@ -86,18 +88,62 @@ def search(initial_state, successor_generator, evaluator_path, environment=None)
...
+
+    message = "No improving successor was found."
+    if report_if_no_improvement:
```
Comment (ClemensBuechner) [inline machetli/search.py:143]: I think the variable name should reflect what we want to do with it. The lines below seem to assume that everything in `report_if_no_improvement` has an `OUT_OF_RESOURCES` status, so the variable should probably be called `tasks_out_of_resources` or something along these lines.

**Review by ClemensBuechner — COMMENTED** [review]
I like the direction we're heading at! The fundamental change of moving the successor choice from the environment to the search makes a lot of sense. I also think the way it is implemented here is pretty clear and understandable, so I'm in favor of continuing down this avenue.

**Comment by FlorianPommerening** [comment]
We decided to leave the handling of out-of-resources errors, moving the evaluator into an executable script, and restarting the search as follow-up issues.

**Pull request merged**